src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+    else
+        h.idxfloor = min(h.idxfloor, index_curr)
+    end
+    return index_curr
+end
+
+#rehash! algorithm
+function rehash!(h::RobinDict{K,V}, newsz = length(h.keys)) where {K, V}
+    oldk = h.keys
+    oldv = h.vals
+    oldh = h.hashes
+    sz = length(oldk)
+    newsz = _tablesz(newsz)
+    h.totalcost += 1
+    if h.count == 0
+        resize!(h.keys, sz)
+        resize!(h.vals, sz)
+        resize!(h.hashes, newsz)
+        fill!(h.hashes, 0)
+        h.count = 0
+        h.maxprobe = 0
+        h.totalcost = 0
+        h.idxfloor = 0
+        return h
+    end
+
+    h.keys = Vector{K}(undef, newsz)
+    h.vals = Vector{V}(undef, newsz)
+    h.hashes = zeros(UInt32,newsz)
+    totalcost0 = h.totalcost
+    h.count = 0
+    h.maxprobe = 0
+    h.idxfloor = 0
+
+    for i = 1:sz
+        @inbounds if oldh[i] != 0
+            k = oldk[i]
+            v = oldv[i]
+            rh_insert!(h, k, v)
+        end
+    end
+    h.totalcost = totalcost0
+    return h
+end
+
+function sizehint!(d::RobinDict, newsz)
+    newsz = _tablesz(newsz*2)  # *2 for keys and values in same array
+    oldsz = length(d.keys)
+    # grow at least 25%
+    if newsz < (oldsz*5)>>2
+        return d
+    end
+    rehash!(d, newsz)
+end
+
+@propagate_inbounds isslotfilled(h::RobinDict, index) = (h.hashes[index] != 0)
+@propagate_inbounds isslotempty(h::RobinDict, index) = (h.hashes[index] == 0)
+
+
+function setindex!(h::RobinDict{K,V}, v0, key0) where {K, V}
+    key = convert(K, key0)
+    isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+    _setindex!(h, key, v0)
+end
+
+function _setindex!(h::RobinDict{K,V}, key::K, v0) where {K, V}
+    v = convert(V, v0)
+    sz = length(h.keys)
+    (h.count > ROBIN_DICT_LOAD_FACTOR * sz) && rehash!(h, sz<<2)
+    index = rh_insert!(h, key, v)
+    @assert index > 0
+    h.totalcost += 1
+    h
+end
+
+isempty(d::RobinDict) = (d.count == 0)
+length(d::RobinDict) = d.count
+
+"""
+    empty!(collection) -> collection
+
+Remove all elements from a `collection`.
+
+# Examples
+```jldoctest
+julia> A = RobinDict("a" => 1, "b" => 2)
+RobinDict{String,Int64} with 2 entries:
+  "b" => 2
+  "a" => 1
+
+julia> empty!(A);
+
+julia> A
+RobinDict{String,Int64} with 0 entries
+```
+"""
+function empty!(h::RobinDict{K,V}) where {K, V}
+    sz = length(h.keys)
+    empty!(h.hashes)
+    empty!(h.keys)
+    empty!(h.vals)
+    resize!(h.keys, sz)
+    resize!(h.vals, sz)
+    resize!(h.hashes, sz)
+    h.count = 0
+    h.maxprobe = 0
+    h.totalcost = 0
+    h.idxfloor = 0
+    return h
+end
+ 
+function rh_search(h::RobinDict{K, V}, key::K) where {K, V}
+    sz = length(h.keys)
+    chash = hash_key(key)
+    index = desired_index(chash, sz)
+    cdibs = 0
+    @inbounds while true
+        if isslotempty(h, index)
+            return -1
+        elseif cdibs > calculate_distance(h, index)
+            return -1
+        elseif h.keys[index] == key && h.hashes[index] == chash
```

Comment:
```suggestion
        elseif h.hashes[index] == chash && (h.keys[index] === key || isequal(h.keys[index], key))
```

I assume you meant to do the hash-check first as a fast path, then the slower equality comparison?